crates/circuit/src/dag_circuit.rs

```diff
@@ -6016,9 +5985,9 @@ impl DAGCircuit {
                         self.vars
                             .find(
                                 &bound_var_map
-                                    .get_item(other.vars.get(*var))?
+                                    .get_item(other.vars.get(*var).unwrap().clone())?
                                     .unwrap()
-                                    .into(),
+                                    .extract()?,
```

Comment:
Do we still need to be using a `PyDict` for this. IIRC the reason I used a pydict here originally was because all this was defined in Python and do to the lookup it was simpler to use python. But now that everything is defined in Rust we should be able to have a hashmap like we do for `Qubit` and `Clbit` right?

Reply:
I just had a look at doing this, and it seems to me like the issue really is going to be that `PyVariableMapper` (which shells out to `imports::VARIABLE_MAPPER`) needs `var_map`, which would of course need to be a Python object.

It looks like it'd be possible for us to port over the internals of `VariableMapper` from Python into Rust at this point, and then we can make that change. Perhaps best for a follow up PR?